Parallel hyperoptimization with MongoDB (#1921)
NNPDF/nnpdf: doc/sphinx/source/n3fit/hyperopt.rst

The above command example is effective when the number of saved trials in the ``test_run/nnfit/replica_1/tries.pkl`` is
less than ``20``. If there are ``20`` or more saved trials, ``n3fit`` will simply terminate, displaying the best results.

Running hyperoptimizations in parallel with MongoDB
---------------------------------------------------

> In NNPDF, you can effectively run hyperoptimization experiments in parallel using `MongoDB <https://www.mongodb.com>`_.

Reviewer: ```suggestion
It is possible to run hyperoptimization scans in parallel using `MongoDB <https://www.mongodb.com>`_.
```

which extends the capabilities of `hyperopt <https://github.com/hyperopt/hyperopt>`_'s `MongoTrials` and enables the
simultaneous evaluation of multiple trials.

To run a parallelized hyperopt search, use the following command:

.. code-block:: bash

  n3fit hyper-quickcard.yml 1 -r N_replicas --hyperopt N_trials --parallel-hyperopt --num-mongo-workers N

Here, ``N`` represents the number of MongoDB workers you wish to launch in parallel.
Each mongo worker handles one trial in Hyperopt. So, launching more workers allows for the simultaneous calculation of a greater number of trials.
Note that there is no need to manually launch MongoDB databases or mongo workers prior to using ``n3fit``,
as the ``mongod`` and ``hyperopt-mongo-worker`` commands are automatically executed
by :meth:`~n3fit.hyper_optimization.mongofiletrials.MongodRunner.start` and
:meth:`~n3fit.hyper_optimization.mongofiletrials.MongoFileTrials.start_mongo_workers` methods, respectivelly.
By default, the ``host`` and ``port`` arguments are set to ``localhost`` and ``27017``, while the database is named ``hyperopt-db``.

> If necessary, you can modify these settings using the ``n3fit --db-host`` , ``n3fit --db-port`` and ``n3fit --db-name`` options.

Reviewer: ```suggestion
If necessary, it is possible modify these settings using the ``n3fit --db-host`` , ``n3fit --db-port`` and ``n3fit --db-name`` options.
```

To resume a hyperopt experiment, add the ``--restart`` option to the ``n3fit`` command:

.. code-block:: bash

  n3fit hyper-quickcard.yml 1 -r N_replicas --hyperopt N_trials --parallel-hyperopt --num-mongo-workers N --restart

Note that, unlike in serial execution, parallel hyperoptimization runs do not generate ``tries.pkl`` files.
Instead, MongoDB databases are saved as ``hyperopt-db.tar.gz`` files inside ``replica_path`` directory.

> These are conveniently extracted for reuse in restart runs.

Reviewer: I am actually wondering if we should either (a) store the history of the database inside the `fit_path` or (b) append to `hyperopt-db` the name of the runcard? The reason is that if one runs different hyperopts with different runcard (with for example different dataset selections) inside a given directory then everything will mix up.

Author: Yes, you are certainly right. we could leave the name of the tar file as general as `hyperopt-db.tar.gz` inside the `fit_path` but compress instead the database name + runcard name ? What do you think ?

Reviewer: A per the discussion, please do whichever is easier (implementation-wise). Option (a) would be nicer but if it turns out to introduce more (unneeded) complications then we should go for (b).

Reviewer 2: We implemented option b. It was failing the hyperopt tests locally, but for seemingly unrelated reasons (I had to install pdfflow?). Let's see if it passes the CI.

Author: The CI/CD will now pass!!

Reviewer: Perfect! Thanks

The above command example is effective when the number of saved trials in the ``test_run/nnfit/replica_1/tries.pkl`` is
less than ``20``. If there are ``20`` or more saved trials, ``n3fit`` will simply terminate, displaying the best results.

Running hyperoptimizations in parallel with MongoDB
---------------------------------------------------

In NNPDF, you can effectively run hyperoptimization experiments in parallel using `MongoDB <https://www.mongodb.com>`_.
This functionality is provided by the :class:`~n3fit.hyper_optimization.mongofiletrials.MongoFileTrials` class,
which extends the capabilities of `hyperopt <https://github.com/hyperopt/hyperopt>`_'s `MongoTrials` and enables the
simultaneous evaluation of multiple trials.

To set up and run a parallelized hyperopt search, follow these steps:

 1. **Initiate the MongoDB database:** Start by setting up the database in your current directory.

>  This database is referred to as ``hyperopt`` in the following instructions. You can initiate it with the command:
> 
>   .. code-block:: bash
> 
>     mongod --dbpath ./hyperopt

Reviewer: Maybe for more clarity? If so, the subsequent occurrences also need to be updated.
```suggestion
 This database is referred to as ``hyperopt-db`` in the following instructions. You can initiate it with the command:

  .. code-block:: bash

    mongod --dbpath ./hyperopt-db
```

  By default, ``mongod`` uses port ``27017``. This is also the default port for the ``n3fit --db-port`` option.

>   If you wish to use a different port, specify it as follows: ``mongod --dbpath ./hyperopt --port YOUR_PORT_NUMBER``.

Reviewer: ```suggestion
  If you wish to use a different port, specify it as follows: ``mongod --dbpath ./hyperopt --db-port YOUR_PORT_NUMBER``.
```
